front/src/modules/trainschedule/components/TimetableV2/useFilterTrainSchedules.ts

```diff
@@ -0,0 +1,112 @@
+import { useEffect, useState } from 'react';
+
+import { uniq } from 'lodash';
+
+import type { ScheduledPointsHonoredFilter, TrainScheduleWithDetails } from './types';
+import { extractTagCode, keepTrain } from './utils';
+
+/**
+ * Hook filtering a train schedules array for timetable train cards
+ * @param trainSchedulesWithDetails the timetable's train schedules
+ * @param debouncedFilter filter on train name and labels
+ * @param debouncedRollingstockFilter filter on train's rolling stock metadata
+ * @param validityFilter filter on valid train or not
+ * @param scheduledPointsHonoredFilter filter on trains that keep their timetables or not
+ * @param selectedTags filter on train's speed limit tag
+ * @returns trainschedules unique speedlimit tags
+ */
+const useFilterTrainSchedules = (
+  trainSchedulesWithDetails: TrainScheduleWithDetails[],
+  debouncedFilter: string,
+  debouncedRollingstockFilter: string,
+  validityFilter: string,
+  scheduledPointsHonoredFilter: ScheduledPointsHonoredFilter,
+  selectedTags: Set<string | null>,
+  setDisplayedTrainSchedules: (trainSchedulesDetails: TrainScheduleWithDetails[]) => void
+) => {
+  const [uniqueTags, setUniqueTags] = useState<string[]>([]);
+
+  const filterTrainSchedules = (trainSchedules: TrainScheduleWithDetails[]) =>
+    trainSchedules.filter((trainSchedule) => {
+      if (!keepTrain(trainSchedule, debouncedFilter)) return false;
+
+      // Apply validity filter
+      if (validityFilter !== 'both') {
+        if (validityFilter === 'valid' && !trainSchedule.isValid) return false;
+        if (validityFilter === 'invalid' && trainSchedule.isValid) return false;
+      }
+
+      // Apply scheduled points honored filter
+      if (scheduledPointsHonoredFilter !== 'both') {
+        if (!trainSchedule.isValid) {
+          return false;
+        }
+        const { scheduledPointsNotHonored } = trainSchedule;
+        if (
+          (scheduledPointsHonoredFilter === 'honored' && scheduledPointsNotHonored) ||
+          (scheduledPointsHonoredFilter === 'notHonored' && !scheduledPointsNotHonored)
+        ) {
+          return false;
+        }
+      }
+
+      // Apply tag filter
+      if (selectedTags.size > 0 && !selectedTags.has(extractTagCode(trainSchedule.speedLimitTag))) {
+        return false;
+      }
+
+      // Apply rolling stock filter
+      if (debouncedRollingstockFilter) {
+        const {
+          detail = '',
+          family = '',
+          reference = '',
+          series = '',
+          subseries = '',
+        } = trainSchedule.rollingStock?.metadata || {};
+        if (
+          ![detail, family, reference, series, subseries].some((v) =>
+            v.toLowerCase().includes(debouncedRollingstockFilter.toLowerCase())
+          )
+        )
+          return false;
+      }
+
+      return true;
+    });
+
+  useEffect(() => {
+    if (trainSchedulesWithDetails.length === 0) {
```

Comment:
I'd seen zero length arrays special-cased elsewhere and held back commenting, but this one is a bit bigger than the rest so I'll say it anyways: I don't think we need the special case? If the regular codepath handles empty arrays just fine, I think the code would be simpler/clearer if we could just drop the `if`.

Reply:
You're right, I removed it :+1: